### PR TITLE
fix(main/dpkg): fixes for the `update-alternatives` -> `mandoc` hook

### PR DESCRIPTION
**Pull request by robertkirkman**
- Fixes https://github.com/termux/termux-packages/issues/25217

- Moves the call to `update_mandoc()` to a codepath that is simultaneously:
  - more likely to run when necessary (such as during `--config` subcommands)
  - and less likely to run when unnecessary (such as during `--install` and `--remove` subcommands)

- I would be interested in test results from others' devices if possible, because unfortunately, when I tested the original version of this patch it initially worked, but then stopped working after a few days.

### CODE REVIEW

**Comment by robertkirkman** [comment]
~~originally, I wanted to figure out a way to make it so that using `update-alternatives --config pager` and pushing Enter without typing anything wouldn't run `makewhatis`, but unfortunately it seems like that's difficult to do without introducing problems in one or another edge case, so this PR's version will always run `makewhatis` with any kind of selection within `update-alternatives --config pager` just in case the alternative has been changed.~~ EDIT: i figured out how to do it, I think.

Subcommands that can't change symbolic links, like `update-alternatives --list pager` and `update-alternatives --get-selections`, still don't run `makewhatis` in this PR's version (for those cases, not running `makewhatis` is the desired behavior)

**Comment by TomJo2000** [comment]
Does the version in this PR run in the `update-alternatives --install` or `--remove` case?

**Comment by robertkirkman** [comment]
> Does the version in this PR run in the `update-alternatives --install` or `--remove` case?

Yes it does, and the version in the original PR also does. I don't know why the original version stopped working when the command `update-alternatives --config pager` is used to change the selection, but in the tests I have done _today_, this PR's version has the effect of making `makewhatis` run _always_ when the `update-alternatives --config` subcommand is used (and `makewhatis` is detected as available), and does not seem to affect anything about the other subcommands, compared to the original PR's version.

**Comment by TomJo2000** [comment]
Could we change the patch to not run in the `--install` or `--remove` codepaths?

I think I would prefer running `makewhatis` as part of the `postinst` and `prerm` scripts instead, so it only runs once per package.
This patch to the build step should accomplish that goal.
```patch
diff --git a/scripts/build/termux_step_update_alternatives.sh b/scripts/build/termux_step_update_alternatives.sh
index bb6b4c977e..60f843de79 100644
--- a/scripts/build/termux_step_update_alternatives.sh
+++ b/scripts/build/termux_step_update_alternatives.sh
@@ -69,9 +69,17 @@ termux_step_update_alternatives() {
 			echo ""
 		done
 		# Close up boilerplate and add end comment
-		echo "  fi"
-		echo "fi"
-		echo "# End automatically added section"
+		printf '%s\n' \
+			"    if [ -x \"${TERMUX_PREFIX}/bin/makewhatis\" ]; then" \
+			"      makewhatis" \
+			"    else" \
+			"      echo Could not find 'makewhatis'," \
+			"      echo cannot update manpage database." \
+			"      echo You can install the mandoc package to fix this." \
+			"    fi" \
+			"  fi" \
+			"fi" \
+			"# End automatically added section"
 		} > postinst
 		if [[ -f postinst.orig ]]; then
 			tail -n+2 postinst.orig >> postinst
@@ -96,9 +104,17 @@ termux_step_update_alternatives() {
 			printf '%s\n' "    update-alternatives --remove \"${name}\" \"${TERMUX_PREFIX}/${ALTERNATIVE[$name]}\""
 		done
 		# Close up boilerplate and add end comment
-		echo "  fi"
-		echo "fi"
-		echo "# End automatically added section"
+		printf '%s\n' \
+			"    if [ -x \"${TERMUX_PREFIX}/bin/makewhatis\" ]; then" \
+			"      makewhatis" \
+			"    else" \
+			"      echo Could not find 'makewhatis'," \
+			"      echo cannot update manpage database." \
+			"      echo You can install the mandoc package to fix this." \
+			"    fi" \
+			"  fi" \
+			"fi" \
+			"# End automatically added section"
 		} > prerm
 		if [[ -f prerm.orig ]]; then
 			tail -n+2 prerm.orig >> prerm
```

**Comment by robertkirkman** [comment]
> Could we change the patch to not run in the `--install` or `--remove` codepaths?

I think that I can make a version that doesn't _necessarily_ run when the `--install` or `--remove` codepaths are used, but does run in any case where that alternative being installed is immediately switched to because of it being the highest priority in its alternative group, which is the specific case of `--install` or `--remove` when it would be necessary.

> I think I would prefer running `makewhatis` as part of the `postinst` and `prerm` scripts instead, so it only runs once per package.

Sure if you prefer this, but keep in mind that if I make the above adjustment, your idea as shown would be less optimal because it would always run `makewhatis` even when the alternative is marked with a lower priority than a previously-installed alternative and isn't being switched to.

**Comment by TomJo2000** [comment]
> I think that I can make a version that doesn't _necessarily_ run when the `--install` or `--remove` codepaths are used, but does run in any case where that alternative being installed is immediately switched to because of it being the highest priority in its alternative group, which is the specific case of `--install` or `--remove` when it would be necessary.

That should work.
I just want to make sure we understand the semantics well.

> Sure if you prefer this, but keep in mind that if I make the above adjustment, your idea as shown would be less optimal because it would always run `makewhatis` even when the alternative is marked with a lower priority than a previously-installed alternative and isn't being switched to.

Yeah that makes sense to me, I think your proposed solution would likely do a better job.
My main concern is reducing unnecessary calls to `makewhatis`.

**Comment by robertkirkman** [comment]
Ok, now I have made and pushed a version that seems to satisfy all requirements.
I have tested this, but I would be interested if you could test it too to look for any additional edge cases that might need to be addressed, since there are a lot of possible configuration states and subcommands for `update-alternatives`.

**Comment by robertkirkman** [comment]
> originally, I wanted to figure out a way to make it so that using `update-alternatives --config pager` and pushing Enter without typing anything wouldn't run `makewhatis`, but unfortunately it seems like that's difficult to do without introducing problems in one or another edge case

I believe this newest version I have made solves this problem (it is a version where `update-alternatives --config pager` and pushing Enter without typing anything doesn't run `makewhatis`, but typing the number of a different selection that isn't the same as the current selection does run `makewhatis`, and simultaneously all other currently-known edge cases seem to be handled correctly)

**Comment by TomJo2000** [comment]
Looks good to me.
Go ahead with the merge when ready.

**Comment by robertkirkman** [comment]
Ok, I will merge it now, and then if you notice any other problems with `update-alternatives`, please ping me ASAP and I will make another PR and/or put changes in the other PR if appropriate:

- https://github.com/termux/termux-packages/pull/25222
